// firestore-storage-core/dist/lib/storage/transformer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_DOCUMENT_TRANSFORMER = exports.ModelClassTransformer = void 0;
const model_1 = require("./model");
const lodash_1 = require("lodash");
class ModelClassTransformer {
    constructor(TypeClass) {
        this.TypeClass = TypeClass;
    }
    fromFirestoreToObject(data, meta) {
        return new this.TypeClass(data, meta);
    }
    toFirestoreDocument(doc) {
        if (doc instanceof model_1.BaseModelClass) {
            return {
                id: doc.id,
                data: doc.getData(),
            };
        }
        return exports.DEFAULT_DOCUMENT_TRANSFORMER.toFirestoreDocument(doc);
    }
}
exports.ModelClassTransformer = ModelClassTransformer;
exports.DEFAULT_DOCUMENT_TRANSFORMER = {
    fromFirestoreToObject(data, meta) {
        const base = {
            id: meta.id,
            _rawPath: meta.rawPath,
        };
        return Object.assign({}, data, base);
    },
    toFirestoreDocument(doc) {
        const clone = (0, lodash_1.cloneDeep)(doc);
        delete clone.id;
        delete clone.createdAt;
        delete clone.updatedAt;
        return {
            id: doc.id,
            data: clone,
        };
    },
};
